flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/OverTypeStrategy.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.utils.LogicalTypeChecks;
+
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/** Specific {@link InputTypeStrategy} for {@link BuiltInFunctionDefinitions#OVER}. */
+public class OverTypeStrategy implements InputTypeStrategy {
+    @Override
+    public ArgumentCount getArgumentCount() {
+        return ConstantArgumentCount.from(4);
+    }
+
+    @Override
+    public Optional<List<DataType>> inferInputTypes(
+            CallContext callContext, boolean throwOnFailure) {
+        final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+        final DataType timeAttribute = argumentDataTypes.get(1);
+
+        if (!LogicalTypeChecks.isTimeAttribute(timeAttribute.getLogicalType())) {
+            return callContext.fail(
+                    throwOnFailure,
+                    "Second argument in OVER window must be a TIME ATTRIBUTE, but is: "
+                            + timeAttribute);
+        }
+
+        if (!callContext.isArgumentLiteral(2)) {
+            return callContext.fail(
+                    throwOnFailure, "Preceding must be a row interval or time interval literal.");
+        }
+
+        if (!callContext.isArgumentLiteral(3)) {
+            return callContext.fail(
+                    throwOnFailure, "Following must be a row interval or time interval literal.");
+        }
+
+        final DataType preceding = argumentDataTypes.get(2);
+        final DataType following = argumentDataTypes.get(3);
+
+        final LogicalTypeRoot precedingTypeRoot = preceding.getLogicalType().getTypeRoot();
+        final Optional<String> precedingErr =
+                validateFollowingPreceding(
+                        precedingTypeRoot, FollowingPreceding.PRECEDING, callContext);
+
+        if (precedingErr.isPresent()) {
+            return callContext.fail(throwOnFailure, precedingErr.get());
+        }
+
+        final LogicalTypeRoot followingTypeRoot = following.getLogicalType().getTypeRoot();
+        final Optional<String> followingErr =
+                validateFollowingPreceding(
+                        followingTypeRoot, FollowingPreceding.FOLLOWING, callContext);
+
+        if (followingErr.isPresent()) {
+            return callContext.fail(throwOnFailure, followingErr.get());
+        }
+
+        final boolean isPrecedingRowWindow =
+                isRowWindow(callContext, precedingTypeRoot, FollowingPreceding.PRECEDING);
+        final boolean isFollowingRowWindow =
+                isRowWindow(callContext, followingTypeRoot, FollowingPreceding.FOLLOWING);
+
+        if (isPrecedingRowWindow && !isFollowingRowWindow
+                || !isPrecedingRowWindow && isFollowingRowWindow) {
+            return callContext.fail(
+                    throwOnFailure, "Preceding and following must be of same interval type.");
```

Comment:
If this is what we are checking, would it be more direct to check the types directly?

Reply:
symbols have symbol types. You can say `2 rows preceding CURRENT_ROW`, but you cannot say `2 rows preceding CURRENT_RANGE`.

Range = time, ROW = number of rows.